usb: USB Audio class implementation

eobalski opened the pull request:
Tested on nRF52840 only, may not work for other platforms.

Two sample apps attached:
-> headphones + microphone
-> headset

Both apps perform loopback. See README.rst for details.

Fixes: #12775
~~Blocked by: #24735~~

carlescufi commented on subsys/usb/class/audio/audio.c:
```diff
@@ -0,0 +1,252 @@
+
+	dev_data = CONTAINER_OF(common, struct usb_audio_dev_data_t, common);
+
+	buffer = net_buf_alloc_fixed(&audio_data_pool, K_NO_WAIT);
```
Use `net_buff_alloc()` instead, it's cleaner.

raveslave commented:
Great to see this work-in-progress! 
- would obviously be nice with support for both audio tx, rx endpoints, format types (stereo, sample rates, clocking etc)
- feature units for controls like mute, gain
- re: Kconfig/dts, I guess we can look a bit on how it's done in linux, there you have switches for pcm, rawmidi and more.

emob-nordic commented:
> * would obviously be nice with support for both audio tx, rx endpoints, format types (stereo, sample rates, clocking etc)

There is plan for this, but I've started with basics.

> * feature units for controls like mute, gain

There is idea to use Kconfig for feature unit associated with headphones (Feature Unit) to choose what should be supported. This [commit](https://github.com/zephyrproject-rtos/zephyr/pull/21526/commits/0e12b7819b5860184edf5b8bf7f2d6a4dda86955) in this PR. This will be reworked in more elegant way using [COND_CODE_1](https://github.com/zephyrproject-rtos/zephyr/blob/master/include/sys/util.h#L320)
> * re: Kconfig/dts, I guess we can look a bit on how it's done in linux, there you have switches for pcm, rawmidi and more.

I will have a look, thanks!

masz-nordic commented on subsys/usb/class/audio/audio.c:
```diff
@@ -199,10 +199,48 @@ void recv_callback(u8_t ep, enum usb_dc_ep_cb_status_code status)
+	dev_data->ops->data_recv_cb(ep, buffer, ret_bytes);
+}
+
+void usb_audio_headphones_register(struct device *dev,
```
Discussed offline:
There will be a  `_register()` function for each common device (headphones, microphone, headset) and one generic. Those will add a device specific union/pointer inside `usb_audio_dev_data_t`.
It would be nice to have a `usb_audio_device_init()` function that is used by all and could handle both predefined structures and user-defined ones in generic case.

masz-nordic commented on include/usb/class/usb_audio.h:
```diff
@@ -0,0 +1,295 @@
+	.bSubframeSize = 2,					\
+	.bBitResolution = 16,					\
+	.bSamFreqType = 1,					\
+	.tSamFreq = {0x80, 0xBB, 0x00}				\
```
This could be set based on `CONFIG_DATA_FORMAT`

masz-nordic commented on subsys/usb/class/audio/Kconfig:
```diff
@@ -0,0 +1,107 @@
+choice
+	prompt "Data format"
+
+config DATA_FORAMT_48KHZ
```
```suggestion
config USB_AUDIO_DATA_FORMAT_48KHZ
```

emob-nordic replied:
When it comes to this. I added this option just for testing purposes. I don't even know if this is relevant for audio.

masz-nordic commented on subsys/usb/class/audio/audio.c:
```diff
@@ -0,0 +1,215 @@
+
+#include <misc/byteorder.h>
+
+/*TODO: Kconfig options for below defines */
```
Just a general note that ISO-specific HW information could be added to device tree.

masz-nordic commented on subsys/usb/class/audio/audio.c:
```diff
@@ -0,0 +1,215 @@
+
+	switch (pSetup->bRequest)
+	{
+	case 0x81:
```
Magic numbers

masz-nordic commented on subsys/usb/class/audio/audio.c:
```diff
@@ -0,0 +1,215 @@
+	}
+}
+
+DEFINE_AUDIO_HEADPHONES_DESCRIPTOR(0, 0x8);
```
This should be first unused iso capable endpoint instead of magic number. Something like `AUTO_EP_IN/OUT` that would also check if the endpoint can handle isochronous transfers.

masz-nordic commented on subsys/usb/usb_device.c:
```diff
@@ -608,15 +608,29 @@ static bool usb_set_interface(u8_t iface, u8_t alt_setting)
+			else if ((cur_iface == iface) && (cur_alt_setting == alt_setting)) {
+				LOG_DBG("Enable endpoint for interface %d, alt setting %d",
+					cur_iface, cur_alt_setting);
+				found = set_endpoint((struct usb_ep_descriptor *)p);
```
Why not simply `usb_dc_ep_enable` when switching to alternate that has already been configured?

emob-nordic replied:
The answer is that when we chose configuration (and associated endpoints) for the first time, there is a need to configure them. When switching to alternate interface there should not be that need. I will think about rework for this.
Endpoints should be configured once and then disabled/enabled depending on chosen alternate interface.

That was left from existing code when I reworked swtiching to alternate interfaces and didn't think about that in this matter.

This is a good point @masz-nordic, thanks! 🚀

Commit:
usb: Sync receiving data from ISO OUT endpoint with SOF

This commit adds support for receiveing data from ISO OUT endpoint
for NRF devices. NRF USB IP core does not generate IRQ when
data are received on ISO OUT endpoint and it must be synchronized
with SOF event.

Enable SOF handling by default if usb audio is configured
with NRF devices.

Signed-off-by: Emil Obalski <emil.obalski@nordicsemi.no>

Commit:
usb: ISO endpoint size may not be power of 2.

By this commit exception for ISO endpoints is made when it comes
to its size. ISO endpoint buffer size for nrf devices is 1023 and
may be configured with variable length size. NRFX checks is size
is chosen accordingly and it is no reason to do it in SHIM.

Signed-off-by: Emil Obalski <emil.obalski@nordicsemi.no>

Commit:
usb: Add Audio class related files

This commit adds files related to USB audio device class.

Signed-off-by: Emil Obalski <emil.obalski@nordicsemi.no>

Commit:
usb: Add USB audio implementation

By this commit USB audio class implementation is introduced
to Zephyr.

The Zephyr USB audio device class follows bellow
documentations:

- Universal Serial Bus specification rev2.0 (usb20.pdf)
- Universal Serial Bus Device Class Definition for Audio Devices
  (audio10.pdf)
- Universal Serial Bus Device Class Definition for Audio Data Formats
  (frmts10.pdf)
- Universal Serial Bus Device Class Definition for Terminal Types
  (termt10.pdf)

Signed-off-by: Emil Obalski <emil.obalski@nordicsemi.no>

Commit:
usb: samples: Add USB audio Headphones + Microphone sample

This commit adds USB audio sample for Zephyr project.

The sample configures two devices:
- Microphone (2 channels, Fs=48kHz, PCM format)
- Headphones (2 channels, Fs=48kHz, PCM format)

Input data stream is passed to output data stream.

For more details refer to README.rst.

Signed-off-by: Emil Obalski <emil.obalski@nordicsemi.no>

jfischer-phytec-iot commented on drivers/usb/device/usb_dc_nrfx.c:
```diff
@@ -1270,6 +1272,35 @@ static void usbd_work_handler(struct k_work *item)
+					usbd_evt_put(iso_ev);
+					usbd_work_schedule();
+				} else {
+					LOG_DBG("Endpoint not ready");
```
`LOG_DBG("Endpoint not enabled");` ?

jfischer-phytec-iot commented on drivers/usb/device/usb_dc_nrfx.c:
```diff
@@ -1270,6 +1272,35 @@ static void usbd_work_handler(struct k_work *item)
 			}
 			break;
 		case USBD_EVT_SOF:
+			if (nrfx_usbd_ep_enable_check(NRFX_USBD_EPOUT8)) {
```
Probably you could move ISO related code to own function e.g. `usbd_sof_trigger_iso_read()`.

jfischer-phytec-iot commented on drivers/usb/device/usb_dc_nrfx.c:
```diff
@@ -1270,6 +1272,35 @@ static void usbd_work_handler(struct k_work *item)
+				}
+				if (ep_ctx->cfg.en) {
+					ep_ctx->read_pending = true;
+					ep_ctx->read_complete = true;
```
That looks weird to set both `read_pending` and `read_complete`

emob-nordic replied:
TBH i do not remember why i did so, probably something was not working, i will check that, thanks.

emob-nordic replied:
OK, I've checked that. I do not know why but when receive request is served (the one I am creating here) both of those flags are checked if they equal 'true' if not, the created event would not be served at all. I did not analyze why it was done like that but it (unluckily) must be here in order to make it work.

jfischer-phytec-iot commented on include/usb/class/usb_audio.h:
```diff
@@ -17,4 +17,573 @@
+#include <net/buf.h>
+#include <sys/util.h>
+
+#if defined(CONFIG_USB_AUDIO_HEADPHONES)
```
L:25 ... L150 seems to be internal configuration dependent stuff, would it not be better to place it in `subsys/usb/audio/audio_internal.h` ?

jfischer-phytec-iot commented on subsys/usb/class/audio/Kconfig:
```diff
@@ -10,4 +10,265 @@ config USB_DEVICE_AUDIO
+module-str = USB Audio
+source "subsys/logging/Kconfig.template.log_config"
+
+menuconfig USB_AUDIO_HEADPHONES
```
Would it make sense to split it in Kconfig, Kconfig.headphones and Kconfig.micro?

emob-nordic replied:
I was thinking about putting all configuration options to *.overlay with relevant binding.

This PR doesn't support configuring 2 pairs of headphones/microphone for now, but I will add it for sure.. This will require having independent configuration options for each pair of Hp/Mic. Having configuration options in Kconfig will require to have fixed number of headphones predefined because Kconfig is not that flexible so it could add configs on the fly. To avoid this I think creating a binding for hp/mic will be nice.

Inside *.overlay user will define as many hps, mics as he wants and the macros needed for configuration will be created automatically.

jfischer-phytec-iot replied:
>Having configuration options in Kconfig will require to have fixed number of headphones predefined because Kconfig is not that flexible so it could add configs on the fly.

I guess something like [_subsys/usb/class/Kconfig.template.composite_device_number_](https://github.com/zephyrproject-rtos/zephyr/blob/master/subsys/usb/class/Kconfig#L30) would help you, please check how it is used for CDC ACM.

emob-nordic replied:
This is not about *_DEVICE_COUNT. The problem is Kconfig will not generate macros like:
*HP_0_FEATURES_MUTE
*HP_1_FEATURES_MUTE
*HP_2_FEATURES_MUTE
depending on *_DEVICE_COUNT == 3

We need to have configuration symbols created on the fly. This is not possible with Kconfig which needs to have those options defined and user is only enabling them.

In other words Kconfig will not add more options if i specify device count 2 or 3. He needs to know it before it is chosen.

jfischer-phytec-iot replied:
You could add a subsys/usb/class/audio/Kconfig.template.audio_hp based on subsys/usb/class/Kconfig.template.composite_device_number, would it not help you?

emob-nordic replied:
If it could work like that then I would need something like this (below) inside Kconfig. Unfortunately this is not possible.
```
switch (CONFIG_USB_AUDIO_HEADPHONES_DEVICE_COUNT)
case 1:
        module = HEADPONES_0
        default-count = 0
        source "subsys/usb/class/audio/Kconfig.template.audio_hp"
case 2:
        module = HEADPONES_0
        default-count = 0
        source "subsys/usb/class/audio/Kconfig.template.audio_hp"
        module = HEADPONES_1
        default-count = 1
        source "subsys/usb/class/audio/Kconfig.template.audio_hp"
.
.
```
where inside subsys/usb/class/audio/Kconfig.template.audio_hp there would be something like:

```
config USB_AUDIO_$(module)_FEATURES_MUTE
        bool "blah blah"
        default y
        help
          Not working
config USB_AUDIO_$(module)_FEATURES_MUTE
.
.
.
```

jfischer-phytec-iot commented on subsys/usb/class/audio/audio.c:
```diff
@@ -10,3 +10,605 @@
+ * @param [in] ep_addr	Endpoint address associated with audio function
+ *
+ */
+#define DEFINE_AUDIO_DESCRIPTOR(dev, link, it_type, ot_type, ep_addr)			\
```
That is not pretty, you could split it to few smaller chunks.

emob-nordic replied:
Will do something like: 
```
#define DEFINE_AUDIO_HP_DESC(dev, link)
        DEFINE_AUDIO_DESCRIPTOR(dev, USB_STREAMING, OUT_HEADPHONES, <addr_from_dt>) 

...

UTIL_LISTIFY(CONFIG_NUM_OF_HP_DEVICES, DEFINE_AUDIO_HP_DESC, _)  //something like this
```
<addr_from_dt> -> see comment in Kconfig

raveslave commented:
Hi Emil,   
Great work! I tested the new sample application and suggest that you change so that the in vs outs are grouped together as "one sound card". Otherwise you'll have issues running both on mac & windows.  
see below:   
![image](https://user-images.githubusercontent.com/1734770/74827104-1a59e280-530d-11ea-9f4e-522a0348ebb2.png)   
   
_(where M4 is a MOTU sound-card)_

jfischer-phytec-iot commented on samples/subsys/usb/audio/src/main.c:
```diff
@@ -0,0 +1,110 @@
+#include <usb/usb_device.h>
+#include <usb/class/usb_audio.h>
+
+#define LOG_LEVEL 3
```
`#define LOG_LEVEL        LOG_LEVEL_INF`

emob-nordic commented:
>  suggest that you change so that the in vs outs are grouped together as "one sound card". Otherwise you'll have issues running both on mac & windows.

Could You please describe the issue You mentioned? That was my intention to be visible as separate devices.

finikorg commented on include/usb/usb_common.h:
```diff
@@ -104,6 +104,7 @@
 	   (USB_CONFIGURATION_ATTRIBUTES_REMOTE_WAKEUP), (0))))
 
 /* Classes */
+#define AUDIO_CLASS			0x01
```
At some point we may rename this and others to `USB_CLASS_AUDIO`, etc

emob-nordic replied:
I've just done it to suit to present defines. I think I will do something about that to resolve #20236 just after I make all corrections to this pr and it will be merged.

finikorg commented on samples/subsys/usb/audio/src/main.c:
```diff
@@ -0,0 +1,110 @@
+#include <usb/class/usb_audio.h>
+
+#define LOG_LEVEL 3
+LOG_MODULE_REGISTER(main);
```
```
LOG_MODULE_REGISTER(main,  LOG_LEVEL_INF);
```
?

finikorg commented on samples/subsys/usb/audio/src/main.c:
```diff
@@ -0,0 +1,110 @@
+		return -EIO;
+	}
+	if (!size) {
+		LOG_ERR("%d", __LINE__);
```
I think we can come up with better log.

emob-nordic replied:
Yea, for sure :) Forgot to remove it after debugging.

finikorg commented on samples/subsys/usb/audio/src/main.c:
```diff
@@ -0,0 +1,110 @@
+
+	net_buf_unref(buffer);
+
+	return 0;
```
Should you return `ret`?  Should you handle `-EAGAIN`

emob-nordic replied:
``` -EAGAIN ``` means the Host did not switchted microphone endpoint to active state -> recording was not started. So the data cannot be send.

Actually this callback should return void. I will update it.

finikorg replied:
You can put this information to comments.

raveslave commented:
Re one device vs two:
When used with a Mac or windows machine, it's most common (and easiest setup for the end user) if it's one unified audio class device. 

It's possible to configure what is called an aggregate device on Mac but it's quite few that knows howto

emob-nordic commented:
> Re one device vs two:
> When used with a Mac or windows machine, it's most common (and easiest setup for the end user) if it's one unified audio class device.
> 
> It's possible to configure what is called an aggregate device on Mac but it's quite few that knows howto

I will check it after I rework it a bit especially when it comes to configuring 2 devices of same type (not supported right now)

Commit:
usb: samples: Add USB Audio Headset sample

This commit adds USB audio sample to the Zephyr project.

The sample configures one I/O device:
- Headset (2 channels, Fs=48kHz, PCM format)

Input data sstream is passed to output data stream.

For more details refer to README.rst.

Signed-off-by: Emil Obalski <emil.obalski@nordicsemi.no>

anangl commented on subsys/usb/class/audio/CMakeLists.txt:
```diff
@@ -0,0 +1,13 @@
+zephyr_library_include_directories(
+  # USB headers
+  ${ZEPHYR_BASE}/include/usb
+  ${ZEPHYR_BASE}/subsys/usb
```
Do you need this actually? See [subsys/usb/CMakeLists.txt](https://github.com/zephyrproject-rtos/zephyr/blob/master/subsys/usb/CMakeLists.txt#L4).

emob-nordic replied:
I used hid and netusb as a reference. I will check if this is still needed.

anangl replied:
Neither of those seems to be needing this entry, as it is already added in the main USB subsys CMake list.
`${ZEPHYR_BASE}/include/usb` seems also to be unnecessary, since you're including this stuff via `#include <usb/*.h>`.

anangl reviewed (commented):
Just a couple of comments for the beginning. I haven't looked gone through the actual implementation yet.

emob-nordic commented on include/usb/class/usb_audio.h:
```diff
@@ -0,0 +1,122 @@
+	usb_audio_feature_updated_cb_t feature_update_cb;
+};
+
+struct controls {
```
What do You think about:
``` diff
/* Feature Unit Control Selectors */
/* audio10.pdf Table A-11 */
-enum feature_unit_control_selectors {
+enum usb_audio_feature_unit_control_selectors {
-	FU_CONTROL_UNDEFINED		= 0x00,
+	USB_AUDIO_FU_CONTROL_UNDEFINED		= 0x00,
-	MUTE_CONTROL			= 0x01,
+	USB_AUDIO_FU_MUTE_CONTROL		= 0x01,
-	VOLUME_CONTROL			= 0x02,
+	USB_AUDIO_FU_VOLUME_CONTROL		= 0x02,
-	BASS_CONTROL			= 0x03,
+	USB_AUDIO_FU_BASS_CONTROL		= 0x03,
-	MID_CONTROL			= 0x04,
+	USB_AUDIO_FU_MID_CONTROL		= 0x04,
-	TREBLE_CONTROL			= 0x05,
+	USB_AUDIO_FU_TREBLE_CONTROL		= 0x05,
-	GRAPHIC_EQUALIZER_CONTROL	= 0x06,
+	USB_AUDIO_FU_GRAPHIC_EQUALIZER_CONTROL	= 0x06,
-	AUTOMATIC_GAIN_CONTROL		= 0x07,
+	USB_AUDIO_FU_AUTOMATIC_GAIN_CONTROL	= 0x07,
-	DELAY_CONTROL			= 0x08,
+	USB_AUDIO_FU_DELAY_CONTROL		= 0x08,
-	BASS_BOOST_CONTROL		= 0x09,
+	USB_AUDIO_FU_BASS_BOOST_CONTROL		= 0x09,
-	LOUDNESS_CONTROL		= 0x0A
+	USB_AUDIO_FU_LOUDNESS_CONTROL		= 0x0A
};
```

When it comes to Feature Unit Control Selectors. For the rest enum/structs I will add ```usb_audio_``` prefix.

emob-nordic commented on samples/subsys/usb/audio/headphones_microphone/src/main.c:
```diff
@@ -0,0 +1,87 @@
+
+	usb_audio_register(hp_dev, &ops);
+
+	usb_audio_register(mic_dev, &ops);
```
Yea, it is. I will create ops for both microphone and headphones seperately.

emob-nordic commented on include/usb/class/usb_audio.h:
```diff
@@ -0,0 +1,122 @@
+ *			inform the user about what is happening
+ */
+void usb_audio_register(struct device *dev,
+			const struct audio_ops *ops);
```
I see this like that: everything is optional and the stack should work even when user does not specify any of the ops. This will be useless but nothing will be broken. Do You think that I should describe it more deeply in the header file?

emob-nordic replied:
@anangl I've provided a more detailed desription for ```struct usb_audio_ops```. I would appreciate if You could have a look.

Commit:
drivers: usb: Correct naming for endpoint Transfer Type

This commit updates definitions to better match usb 2.0 spec.

Signed-off-by: Emil Obalski <emil.obalski@nordicsemi.no>

emob-nordic commented:
Found a bug in hp+mic sample when running on linux, working on fix. Will remain DNM until repaired.

emob-nordic commented:
* updated with possibility to choose between Endpoint Synchornization Type
* fixed a critical issue with Linux environment (Host invokes request with always interface_id==0)
* rebased onto current master

emob-nordic commented on drivers/usb/device/usb_dc_nrfx.c:
```diff
@@ -1231,7 +1231,48 @@ static inline void usbd_reinit(void)
+			LOG_DBG("Endpoint is not enabled");
+		}
+	} else {
+		/* Do nothing */
```
Maybe:
``` diff
 /**
  * @brief funciton to generate fake receive request for
  * ISO OUT EP.
@@ -1244,33 +1245,30 @@ static void usbd_sof_trigger_iso_read(void)
        struct usbd_event *ev;
        struct nrf_usbd_ep_ctx *ep_ctx;
 
-       if (nrfx_usbd_ep_enable_check(NRFX_USBD_EPOUT8)) {
-               /** Dissect receive request
+       ep_ctx = endpoint_ctx(NRFX_USBD_EPOUT8);
+       if (!ep_ctx) {
+               LOG_ERR("There is no ISO ep");
+               return;
+       }
+
+       if (nrfx_usbd_ep_enable_check(NRFX_USBD_EPOUT8) && ep_ctx->cfg.en) {
+               /* Dissect receive request
                 * if the iso OUT ep is enabled
                 */
-               ep_ctx = endpoint_ctx(NRFX_USBD_EPOUT8);
-               if (!ep_ctx) {
-                       LOG_ERR("There is not ISO ep");
+               ep_ctx->read_pending = true;
+               ep_ctx->read_complete = true;
+               ev = usbd_evt_alloc();
+               if (!ev) {
+                       LOG_ERR("Failed to alloc evt");
                        return;
                }
-               if (ep_ctx->cfg.en) {
-                       ep_ctx->read_pending = true;
-                       ep_ctx->read_complete = true;
-                       ev = usbd_evt_alloc();
-                       if (!ev) {
-                               LOG_ERR("Failed to alloc evt");
-                               return;
-                       }
-                       ev->evt_type = USBD_EVT_EP;
-                       ev->evt.ep_evt.evt_type = EP_EVT_RECV_REQ;
-                       ev->evt.ep_evt.ep = ep_ctx;
-                       usbd_evt_put(ev);
-                       usbd_work_schedule();
-               } else {
-                       LOG_DBG("Endpoint is not enabled");
-               }
+               ev->evt_type = USBD_EVT_EP;
+               ev->evt.ep_evt.evt_type = EP_EVT_RECV_REQ;
+               ev->evt.ep_evt.ep = ep_ctx;
+               usbd_evt_put(ev);
+               usbd_work_schedule();
        } else {
-               /* Do nothing */
+               LOG_DBG("Endpoint is not enabled");
        }
 }
```

emob-nordic commented on drivers/usb/device/usb_dc_nrfx.c:
```diff
@@ -1231,7 +1231,48 @@ static inline void usbd_reinit(void)
+	struct nrf_usbd_ep_ctx *ep_ctx;
 
+	if (nrfx_usbd_ep_enable_check(NRFX_USBD_EPOUT8)) {
+		/** Dissect receive request
```
No. I will remove one * this will I believe leave single comment?

carlescufi commented:
@galak I would like to include this PR for tomorrow's dev review meeting.

The issue we'd like to discuss is the usage of Devicetree to define the properties of multiple instances of the USB audio class.

The issue faced with this PR is that the user can define any number of instances of the following audio device types:

- headphones
- microphone
- headset

Each of them needs a set of parameters, and Kconfig doesn't allow looping based on a template, so the only solution found was to use Device Tree:

- Example of binding https://github.com/zephyrproject-rtos/zephyr/pull/21526/commits/b37689a540ed46aa500743e543f548d6d78ff4aa#diff-2bc8474e40d7fb53547a1950bdbfdb7cR1
- Example of dts file: https://github.com/zephyrproject-rtos/zephyr/pull/21526/files#diff-32f877b98f3ca2ca3a1fde4999344208R7

Right now the counts of each audio device type also need to be defined in Kconfig:
- https://github.com/zephyrproject-rtos/zephyr/pull/21526/commits/b37689a540ed46aa500743e543f548d6d78ff4aa#diff-0dfd9e9e33f89277b14a9b32cae58478R30, but this should be replaced with DT macros instead, since DT has the info about the number of instances of each audio device type

The question we'd like to ask is whether using DT in this way is acceptable, or should we instead use Kconfig with a maximum number of instances.

arvid-r commented:
Great work! I'll try to test this too on one of the NXP iMX boards. Just like @raveslave mentions, it would be very useful with support for just generic devices that are not headphones or microphones, and have both inputs and outputs at the same time.

galak commented:
Dev-review (Apr 16): General feeling was that DT was reasonable from at least those in the meeting.  Should support all classes configured by DT at some point so the same mechanism is used.

jfischer-phytec-iot commented on samples/subsys/usb/audio/headphones_microphone/nrf52840dk_nrf52840.overlay:
```diff
@@ -0,0 +1,25 @@
+		channel-l;
+		channel-r;
+		resolution = <16>;
+		ep-addr = <0x08>;
```
Please remove `ep-addr` and `ep-sync-type`, they should not be there and could lead to assumption it can be configured.

emob-nordic replied:
When it comes to ep-addr I will remove it and make it work with AUTO_EP_IN/OUT.
When it comes to ep-sync-type it actually can be configured, it's information to HOST how to synchronize the clocks.
Anyway I will delete this as Synchronization type Asynchronous for OUT endpoint requires feedback ISO IN endpoint and we will not support this for now.

emob-nordic replied:
The default configuration will remain Synchronous as current descriptor structure supports this configuration.

emob-nordic replied:
@jfischer-phytec-iot updated.

arvid-r commented:
Oh, I see, the headset mode might be useful for our use case. I will try it out on my board. I just wished there was a more generic case for X number of inputs and Y number of outputs not being referred to as "Center", "Left" etc. but this might be a cosmetic detail I guess.

jfischer-phytec-iot commented on samples/subsys/usb/audio/headphones_microphone/nrf52840dk_nrf52840.overlay:
```diff
@@ -0,0 +1,25 @@
+/*
```
please rename nrf52840dk_nrf52840.overlay to app.overlay, the sample has to work on any system with isochronous endpoints.

emob-nordic replied:
Done.

emob-nordic commented on include/usb/class/usb_audio.h:
```diff
@@ -17,4 +17,135 @@
+	usb_audio_feature_updated_cb_t feature_update_cb;
+};
+
+struct usb_audio_controls {
```
I think it should. User may want to use this to trace the current state of the feature unit especially when it comes to mute and volume.

emob-nordic commented on subsys/usb/class/audio/Kconfig:
```diff
@@ -0,0 +1,42 @@
+	default "AUDIO"
+	help
+	  Device name template for the AUDIO Devices.
+	  Headphones device would have name
+	  $(USB_AUDIO_DEVICE_NAME)_HP_0.
+	  Microphone device would have name
+	  $(USB_AUDIO_DEVICE_NAME)_MIC_0.
+	  Headset device would have name
+	  $(USB_AUDIO_DEVICE_NAME)_HS_0.
```
Yes. I am on the way to move everything what is left in Kconfig to dt as we agreed in dev-review.

emob-nordic replied:
Done.

Commit:
drivers: usb: Add enum type for Synchronization Type.

This commit adds enumeration type for isochronous
endpoints specific information required by USB Audio Class.

Signed-off-by: Emil Obalski <emil.obalski@nordicsemi.no>

emob-nordic commented on samples/subsys/usb/audio/headset/README.rst:
```diff
@@ -0,0 +1,36 @@
+********************
+
+This sample can be built for multiple boards. In order to build for boards different than
+nrf52840_pca10056 proper overlay file must be provided.
```
The description is updated @anangl please have a look.

emob-nordic commented:
Pushed the branch with following changes:
-

- DEVICE_NAME option moved from Kconfig to DT.
- DEVICE_COUNT option moved from Kconfig to DT (possibility to configure 2 devices of given type).
- ep-sync-type removed and fixed to be Synchronous.
- add Synchronization Type enum for ISO needs. (seperate commit).
- Some enumeration types brought back to public API.
- Sample description updated.
- Minor cleanup for descriptors.

emob-nordic commented:
> Oh, I see, the headset mode might be useful for our use case. I will try it out on my board. I just wished there was a more generic case for X number of inputs and Y number of outputs not being referred to as "Center", "Left" etc. but this might be a cosmetic detail I guess.

Believe me it's not cosmetic. "Center" and "Left" are channels which are part of single pipe crated between device EP and host. Additional Input/Output changes the descriptors dramatically (it adds additional EP which is a part of same device - the descriptor grows). Adding channels does not change the descriptor structure.

arvid-r commented:
> > Oh, I see, the headset mode might be useful for our use case. I will try it out on my board. I just wished there was a more generic case for X number of inputs and Y number of outputs not being referred to as "Center", "Left" etc. but this might be a cosmetic detail I guess.
> 
> Believe me it's not cosmetic. "Center" and "Left" are channels which are part of single pipe crated between device EP and host. Additional Input/Output changes the descriptors dramatically (it adds additional EP which is a part of same device - the descriptor grows). Adding channels does not change the descriptor structure.
> @arvid-r

Really? I still don't know enough about this obviously, but I'll try to look into these things soon. At this point I just hope that it will be feasible to add support for more generic devices in the future, which we most likely can spend some time on.

emob-nordic commented:
Pushed the branch with following changes:
-

- EP_SIZE now rely on device configuration (ch. num, bit resolution, s.freq)
- net_buf is created per OUT device, user responsible for managing buffer for IN devices
- deleted buffer allocator as not required now
- mic/hp part in headset can now have totally seperate configuration
- concatenated some INIT macros into one.
- API documentation updated to above changes.

emob-nordic commented:
Pushed the branch with following changes:
-

- Configuring synchronization type for IN endpoints is back possible.
- Repaired checkpatch errors.
- squashed commit adding composite support into implementation commit.

arvid-r commented:
First try on iMX RT 1020 evk. Output seems to be working, but I get only zeros in when trying the headset example. Will dig some more.

anangl commented on include/usb/class/usb_audio.h:
```diff
@@ -0,0 +1,286 @@
+ * @param evt Event information to be parsed by app.
+ */
+typedef void (*usb_audio_feature_updated_cb_t)(
+					struct usb_audio_feature_unit_evt evt);
```
For consistency, it would be better to pass the `dev` pointer in the first parameter of the callback function, not as a field of this structure.

anangl commented on include/usb/class/usb_audio.h:
```diff
@@ -0,0 +1,286 @@
+	 */
+	usb_audio_data_completion_cb_t data_received_cb;
+
+	/* Callback called when features were manipulation by Host */
```
manipulation?

emob-nordic replied:
Whoops.. will reword this

anangl commented on include/usb/class/usb_audio.h:
```diff
@@ -0,0 +1,286 @@
+ *	  This must be called in order to make the device work
+ *	  and respond to all relevant requests.
+ *
+ * @param dev USB audio device which will send the data
+ *	      over its ISO IN endpoint
```
And what about OUT only devices?

emob-nordic replied:
I must have forgotten to update it, its copied from ```usb_audio_send```..

emob-nordic commented on subsys/usb/class/audio/audio.c:
```diff
@@ -0,0 +1,1023 @@
+#include <logging/log.h>
+LOG_MODULE_REGISTER(usb_audio, CONFIG_USB_AUDIO_LOG_LEVEL);
+
+static void audio_buffer_destroyed(struct net_buf *buf);
```
Will get rid of it.

emob-nordic commented on subsys/usb/class/audio/audio.c:
```diff
@@ -0,0 +1,1023 @@
+		&usb_audio_data_devlist);
+}
+
+static void audio_buffer_destroyed(struct net_buf *buf)
+{
+	net_buf_destroy(buf);
+}
```
s/a

emob-nordic commented on subsys/usb/class/audio/audio.c:
```diff
@@ -0,0 +1,1023 @@
+	return 0;
+}
+
+static const struct usb_audio_device_api {
+	void (*init)(void);
+} usb_audio_api;
```
Relic. Will remove this.

emob-nordic commented on subsys/usb/class/audio/audio.c:
```diff
@@ -0,0 +1,1023 @@
+ * @param [in] it_type	Input terminal type
+ * @param [in] ot_type	Output terminal type
+ */
+#ifdef CONFIG_USB_COMPOSITE_DEVICE
+#define DEFINE_AUDIO_DESCRIPTOR(dev, i, id, link, it_type, ot_type,	      \
+				sync_type, addr)			      \
+USBD_CLASS_DESCR_DEFINE(primary, audio)					      \
+struct dev##_descriptor_##i dev##_desc_##i = {				      \
+	.iad = INIT_IAD(USB_AUDIO_AUDIOCONTROL),			      \
```
Great idea. I will try this out.

emob-nordic commented on subsys/usb/class/audio/audio.c:
```diff
@@ -0,0 +1,1023 @@
+#define DEFINE_AUDIO_HS_DEVICE_AUTO(i, _) \
+	DEFINE_AUDIO_DEVICE(HS, i)
+
+UTIL_LISTIFY(HEADPHONES_DEVICE_COUNT,
+		DECLARE_AUDIO_HP_DESCR_AUTO, _)
+UTIL_LISTIFY(HEADPHONES_DEVICE_COUNT,
+		DEFINE_AUDIO_HP_DESCR_AUTO, _)
+UTIL_LISTIFY(HEADPHONES_DEVICE_COUNT,
+		DEFINE_AUDIO_HP_EP_AUTO, _)
+UTIL_LISTIFY(HEADPHONES_DEVICE_COUNT,
+		DEFINE_AUDIO_HP_CFG_DATA_AUTO, _)
+UTIL_LISTIFY(HEADPHONES_DEVICE_COUNT,
+		DEFINE_AUDIO_HP_POOL_AUTO, _)
+UTIL_LISTIFY(HEADPHONES_DEVICE_COUNT,
+		DEFINE_AUDIO_HP_DEV_DATA_AUTO, _)
+UTIL_LISTIFY(HEADPHONES_DEVICE_COUNT,
+		DEFINE_AUDIO_HP_DEVICE_AUTO, _)
```
Need to check that, will give it a try. Haven't considered this before.

anangl commented on subsys/usb/class/audio/audio.c:
```diff
@@ -0,0 +1,1023 @@
+	struct feature_unit_descriptor *fu = NULL;
+	struct usb_audio_feature_unit_evt evt = { .dev = audio_dev_data->common.dev };
+	u8_t ch_num, ch_start, ch_end;
+	static u8_t tmp_data_ptr[3];
```
This assumes that there at most 3 channels, and each control uses an `u8_t` value. Is that guaranteed some way?

emob-nordic replied:
It's not. I forgot about that. Will think about rework.

anangl commented on subsys/usb/class/audio/audio.c:
```diff
@@ -0,0 +1,1023 @@
+	struct feature_unit_descriptor *fu;
+
+	header = audio_dev_data->header_descr;
+	fu = (struct feature_unit_descriptor *)((u8_t *)header +
```
When a value is to be only read, it is better to use `const` pointer to it.

anangl commented on subsys/usb/class/audio/audio.c:
```diff
@@ -0,0 +1,1023 @@
+	}
+}
+
+void audio_cb_usb_status(struct usb_cfg_data *cfg,
```
`static` missing.

anangl commented on subsys/usb/class/audio/audio.c:
```diff
@@ -0,0 +1,1023 @@
+	return 1;
+}
+
+void audio_dc_interface(struct usb_audio_dev_data_t *audio_dev_data,
```
`static` missing.

anangl commented on subsys/usb/class/audio/audio.c:
```diff
@@ -0,0 +1,1023 @@
+ * This is needed in order to address audio specific requests to proper
+ * controls struct.
+ */
+int get_feature_unit(struct usb_audio_dev_data_t *audio_dev_data,
```
`static` missing.

anangl commented on subsys/usb/class/audio/audio.c:
```diff
@@ -0,0 +1,1023 @@
+/**
+ * Helper function for getting the device streaming direction
+ */
+enum usb_audio_direction get_fu_dir(struct feature_unit_descriptor *fu)
```
`static` missing.

anangl commented on include/usb/class/usb_audio.h:
```diff
@@ -0,0 +1,286 @@
+ * This structure present the structure of possible features that USB audio
+ * device may support. Enabling features is done via DT overlay.
+ */
+struct usb_audio_features {
```
Why is this structure defined in the public API? Is the application supposed to somehow access it?

emob-nordic replied:
Hmm, no. I will put this to the internals.

anangl commented on subsys/usb/class/audio/audio.c:
```diff
@@ -0,0 +1,1023 @@
+	};
+
+#define DEFINE_AUDIO_DEV_DATA(dev, i, __out_pool, __in_pool_size)	\
+	struct usb_audio_features dev##_ctrls_##i[CH_CNT(dev, i) + 1];	\
```
`static` missing, same story in `DEFINE_AUDIO_DEV_DATA_BIDIR`.

emob-nordic commented on subsys/usb/class/audio/usb_audio_internal.h:
```diff
@@ -0,0 +1,506 @@
+#define MIC_LINK(i) ((3*(HEADPHONES_DEVICE_COUNT + i)) + 3)
+
+#define FEATURES(dev, i) (0x0000					     \
+| COND_CODE_1(DT_INST_##i##_USB_AUDIO_##dev##_FEATURE_MUTE,		     \
+	(BIT(0)), (0))							     \
```
Tbh I am not quite convinced if this is actually a simplification. The only advantage from that is properties for FEATURES(dev, i) macro are defined in one line. One disadvantage is that its not visible explicitly that FEATURES/CH_CFG does come from DT. Let me know what You think I do not insist I could change that as this is more elegant but harder to guess where the options comes from.

emob-nordic replied:
@jfischer-phytec-iot any opinion?

emob-nordic commented on subsys/usb/class/audio/usb_audio_internal.h:
```diff
@@ -0,0 +1,506 @@
+}
+
+/** Class-Specific AS Format Type Descriptor 4.5.3 audio10.pdf
+ * TODO: bFormatType
```
I will remove this todo. I left this from the very beggining when I thaugh it would be easy to swap between formats (there are 3 in total). We are supporting format type I so it must be `0x01` and nothing more.

emob-nordic commented on subsys/usb/class/audio/audio.c:
```diff
@@ -0,0 +1,1023 @@
+		net_buf_unref(buffer);
+		return;
+	}
+	if (audio_dev_data->ops && audio_dev_data->ops->data_received_cb) {
+		audio_dev_data->ops->data_received_cb(common->dev, buffer, ret_bytes);
+	}
+}
```
Thats right.. I assumed the user will always define received_cb as without that the device is useless BUT this is an issue indeed. I will change that, thanks!

emob-nordic commented on subsys/usb/class/audio/audio.c:
```diff
@@ -0,0 +1,1023 @@
+	return audio_dev_data->in_frame_size;
+}
+
+void audio_receive_cb(u8_t ep, enum usb_dc_ep_cb_status_code status)
```
There are many places I forgot to add `static`.. I will update this.

emob-nordic commented on subsys/usb/class/audio/audio.c:
```diff
@@ -0,0 +1,1023 @@
+
+	/* Call user callback if user registered one */
+	if (audio_dev_data->ops && audio_dev_data->ops->data_written_cb) {
+		audio_dev_data->ops->data_written_cb(dev_data->dev, NULL, size);
```
I will reword the API. Same callback type is used for send/received there is where the issue comes from. I will distinguish the behavior when it comes to the buffer.
I think there is no any reason so that the user must free the buffer after send so this will remain NULL.

emob-nordic replied:
Discussed offline and changed my mind. Will pass the buffer and depending on user app (defined the callback or not). The user is responsible/not responsible for freeing the buffer. This is more consistent approach.

emob-nordic commented on subsys/usb/class/audio/audio.c:
```diff
@@ -0,0 +1,1023 @@
+
+	switch (REQTYPE_GET_RECIP(pSetup->bmRequestType)) {
+	case REQTYPE_RECIP_INTERFACE:
+		if (handle_interface_req(pSetup, len, data) < 0) {
+			return -1;
```
Yes thats right. I dont know why I didn't do it this way... Will update.

emob-nordic commented on subsys/usb/class/audio/audio.c:
```diff
@@ -0,0 +1,1023 @@
+		}
+		break;
+	case REQTYPE_RECIP_ENDPOINT:
+		if (handle_endpoint_req(pSetup, len, data) < 0) {
```
I wanted to stay close to the usb audio spec. It defines endpoint specific requests as well (which we do not support by this implementation). I wanted to leave this open for the discussion.
Compiler will get rid of it for sure. The question is if we want to underline existance of such requests and if its relevant to leave this handler for future?

emob-nordic replied:
Removed.

emob-nordic commented on subsys/usb/class/audio/audio.c:
```diff
@@ -0,0 +1,1023 @@
+		break;
+	default:
+		LOG_ERR("Request receipent invalid");
+		return -1;
```
It failed. The request is wrong and we should return negative. `REQTYPE_GET_RECIP()` may return `REQTYPE_RECIP_DEVICE` which is considered wrong and should return error.

anangl replied:
I agree, it failed. The point is that it should return some negative errno code, not just -1.

emob-nordic replied:
Ah, okay. `-EINVAL`?

emob-nordic replied:
Replaced with `-EINVAL`

emob-nordic commented on subsys/usb/class/audio/audio.c:
```diff
@@ -0,0 +1,1023 @@
+		case USB_AUDIO_FU_MUTE_CONTROL:
+			switch (pSetup->bRequest) {
+			case USB_AUDIO_SET_CUR:
+				memcpy(&audio_dev_data->controls[device][ch].mute,
+					(*data + data_offset),
+					sizeof(bool));
```
Thats, right. Will update the code.

emob-nordic commented on subsys/usb/class/audio/audio.c:
```diff
@@ -0,0 +1,1023 @@
+			}
+			data_offset++;
+			break;
+		case USB_AUDIO_FU_VOLUME_CONTROL:
+		case USB_AUDIO_FU_BASS_CONTROL:
+		case USB_AUDIO_FU_MID_CONTROL:
+		case USB_AUDIO_FU_TREBLE_CONTROL:
+		case USB_AUDIO_FU_GRAPHIC_EQUALIZER_CONTROL:
+		case USB_AUDIO_FU_AUTOMATIC_GAIN_CONTROL:
+		case USB_AUDIO_FU_DELAY_CONTROL:
+		case USB_AUDIO_FU_BASS_BOOST_CONTROL:
+		case USB_AUDIO_FU_LOUDNESS_CONTROL:
+			break;
```
I left this for the future as by this PR I am only supporting mute feature and nothing more. I wanted to underline there that additional support could be implemented. Mute is core feature that Host requires for the  device to work.

It could return last read mute values only if user chose property in the DT. It's explicitly said there that those are currently not supported.
To avoid this I instead of `break;` i can put `return -ENOTSUP`

emob-nordic replied:
Removed unused cases.

emob-nordic commented:
Pushed the branch with updates addresing @anangl comments.
Branch is now dependent on #24735.
Branch is presenting the fix for #24200 which is global, class specific issue.
This PR does not solve the issue absolutely.

anangl commented on subsys/usb/class/audio/audio.c:
```diff
@@ -0,0 +1,958 @@
+{
+	u8_t ep_addr;
+
+	for (int i = 0; i < cfg->num_endpoints; i++) {
+		ep_addr = cfg->endpoint[0].ep_addr;
```
Something is probably wrong here. There is a loop but in each iteration it only checks the first endpoint.

emob-nordic replied:
Ah, yea... Initial thought was it should loop over endpoints but then I made IN endpoints always 1st. I will change this -> remove the loop.

anangl commented on subsys/usb/class/audio/audio.c:
```diff
@@ -0,0 +1,958 @@
+ */
+static enum usb_audio_direction get_fu_dir(struct feature_unit_descriptor *fu)
+{
+	struct output_terminal_descriptor *ot  =
+	(struct output_terminal_descriptor *)((u8_t *)fu + fu->bLength);
```
Imperfect indentation. And `const`s missing (in the function parameter as well).

anangl commented on subsys/usb/class/audio/audio.c:
```diff
@@ -0,0 +1,958 @@
+	case USB_AUDIO_FU_MUTE_CONTROL:
+		switch (pSetup->bRequest) {
+		case USB_AUDIO_SET_CUR:
+			if (ch == 0xFF) {
+				memcpy(&audio_dev_data->controls[device][POS(MUTE, 0, audio_dev_data->ch_num[device])],
+				       *data, LEN(audio_dev_data->ch_num[device], MUTE));
+			} else {
+				audio_dev_data->controls[device][ch] =
+							(*data[0] != 0);
+			}
```
Why not use local variables:
```
	u8_t *controls = audio_dev_data->controls[device];
	u8_t ch_num = audio_dev_data->ch_num[device];
```
and shorten the above to:
```
			if (ch == 0xFF) {
				memcpy(&controls[POS(MUTE, 0, ch_num)],
				       *data, LEN(ch_num, MUTE));
			} else {
				controls[POS(MUTE, ch, ch_num)] = (*data)[ch];
			}
```
?
BTW `*data[0]` is wrong in this context. First of all it always gets the value of the first channel instead of the channel `ch`. But to avoid confusion, `*data` should be in parentheses because `*data[i]` here is equivalent to `(*data)[i]` only when `i` is 0.
And don't you need to bother with `*len` here?

emob-nordic replied:
> BTW *data[0] is wrong in this context. First of all it always gets the value of the first channel instead of the channel ch.

It does not. If `ch != 0xFF` the data for addressed channel are stored starting from index 0 of data pointer. For example if `ch = 2` then data are under `(*data)[0]` . The len should remain '1' (in case of MUTE) that's why I omitted this but worthy to bring it back.

emob-nordic replied:
I will use locals for sure, will make the code much easier to read.

anangl replied:
> > First of all it always gets the value of the first channel instead of the channel ch.
> 
> It does not. If ch != 0xFF the data for addressed channel are stored starting from index 0 of data pointer.

Right. My mistake. But why use an index then when the pointed value is not an array?

emob-nordic replied:
Uhm.. my bad, will change to `**data` as it should be from the beginning.

anangl commented on subsys/usb/class/audio/audio.c:
```diff
@@ -0,0 +1,958 @@
+				evt.cs = cs;
+				evt.channel = ch;
+				evt.dir = get_fu_dir(fu);
+				evt.val = &audio_dev_data->controls[device][POS(MUTE, 0, audio_dev_data->ch_num[device])];
+				evt.val_len = (ch == 0xFF) ? LEN(audio_dev_data->ch_num[device], MUTE) : LEN(1, MUTE);
```
When `ch != 0xFF`, the pointed value should be for the channel `ch`, not always the first one, shouldn't it?

emob-nordic replied:
Right, my mistake.

emob-nordic commented on include/usb/class/usb_audio.h:
```diff
@@ -0,0 +1,282 @@
+ *
+ * @param dev	 The device for which the callback was called.
+ * @param buffer Pointer to the net_buf data chunk that was successfully
+ *		 send/received. If application does not specify the callback
+ *		 the buffer would be automatically freed.
```
` Pointer to the net_buf data chunk that was successfully send/received. If the application uses data_written_cb and/or data_received_cb callbacks it is responsible for freeing the buffer by itself.` ?

emob-nordic commented on include/usb/class/usb_audio.h:
```diff
@@ -0,0 +1,282 @@
+ * @param buffer Pointer to the buffer that should be send. User is
+ *		 responsible for managing the buffer for Input devices.
+ *		 In case of sending error user must decide if buffer
+ *		 should be dropped or retransmitted.
```
``` Pointer to the buffer that should be send. User is responsible for managing the buffer for Input devices. In case of sending error user must decide if the buffer should be dropped or retransmitted. Afther the buffer was sent successfully it is passed to the data_written_cb callback if the application uses one or automatically freed otherwse. User must provide proper net_buf chunk especially when it comes to its size. This information can be obtained using usb_audio_get_in_frame_size() API function. ```

emob-nordic commented on samples/subsys/usb/audio/headphones_microphone/prj.conf:
```diff
@@ -0,0 +1,21 @@
+
+#net buf options
+CONFIG_NET_BUF=y
+CONFIG_NET_BUF_USER_DATA_SIZE=4
```
Not used. I forgot about this. By default it is 4 but as we do not use it I think it's worth to be configured 0.

jfischer-phytec-iot commented on subsys/usb/class/audio/usb_audio_internal.h:
```diff
@@ -0,0 +1,527 @@
+ * Actual size of the struct should be checked by reading
+ * .bLength.
+ */
+struct cs_ac_interface_descriptor_header {
```
please use shorter names, this one is 33 characters long, you can use _iface_ for interface and _desc_ for descriptor as abbreviation

emob-nordic replied:
I will replace it with `cs_ac_if_descriptor` to be more or less consistent with what we have in the tree, see [usb_if_descriptor](https://github.com/zephyrproject-rtos/zephyr/blob/master/include/usb/usb_common.h#L188).

jfischer-phytec-iot commented on subsys/usb/class/audio/audio.c:
```diff
@@ -10,3 +10,952 @@
+
+	size_t ch_num[2];
+
+	const struct cs_ac_interface_descriptor_header *header_descr;
```
*desc_header or *desc_hdr

jfischer-phytec-iot commented on subsys/usb/class/audio/audio.c:
```diff
@@ -10,3 +10,952 @@
+LOG_MODULE_REGISTER(usb_audio, CONFIG_USB_AUDIO_LOG_LEVEL);
+
+/* Device data structure */
+struct usb_audio_dev_data_t {
```
`struct usb_audio_dev_data` or shorter `usb_audio_data` or `audio_context`

emob-nordic replied:
Will go for `usb_audio_data` as I want to underline this is audio specific structure.

emob-nordic replied:
Changed my mind, I think better chose would be `struct usb_audio_dev_data` as simple `usb_audio_data` may suggest its data packet or whatsoever. `dev` addition clarifies a lot.

jfischer-phytec-iot commented on include/usb/class/usb_audio.h:
```diff
@@ -10,11 +10,277 @@
+/** Feature Unit Control Selectors
+ * Refer to Table A-11 from audio10.pdf
+ */
+enum usb_audio_feature_unit_control_selectors {
```
you have to find shorter name for it and probably other enums here, maybe `usb_audio_fucs`?

emob-nordic replied:
Hmm. Wanted to underline that comes 1:1 from the spec but as there is a comment about this its worth to be shorten.

jfischer-phytec-iot commented on subsys/usb/class/audio/audio.c:
```diff
@@ -10,3 +10,952 @@
+ *
+ * @return audio_dev_data for given interface, NULL if not found.
+ */
+static struct usb_audio_dev_data_t *get_audio_dev_data_by_interface(
```
as this is static, why not use shorter names without prefix like `static struct audio_context *get_dev_data_by_iface`

emob-nordic replied:
For sure not `get_dev_data_by_iface` as `usb_get_dev_data_by_iface` is used in the usb_descriptor.h and this could be misleading. Will replace with:
`static struct usb_audio_data *get_audio_data_by_iface()`

emob-nordic replied:
According to previous comment: `get_audio_dev_data_by_iface()`

jfischer-phytec-iot commented on subsys/usb/class/audio/audio.c:
```diff
@@ -10,3 +10,952 @@
+								u8_t interface)
+{
+	struct usb_dev_data *dev_data;
+	struct usb_audio_dev_data_t *audio_dev_data;
```
`struct audio_context *ctx;` ?

emob-nordic replied:
```struct usb_audio_data *audio_data;```

emob-nordic replied:
`struct usb_audio_dev_data *audio_dev_data;`

jfischer-phytec-iot commented on subsys/usb/class/audio/audio.c:
```diff
@@ -10,3 +10,952 @@
+	}
+
+	switch (cs) {
+	case USB_AUDIO_FU_MUTE_CONTROL:
```
```c
if (cs != USB_AUDIO_FU_MUTE_CONTROL) {
    return -ENOSUP;
}
```
You will probably expand it in the future but then you can move MUTE_CONTROL to a handle_fu_req_mute function.

emob-nordic replied:
I will expand and yes I planned to create something like `handle_fu_req_mute()` in the future. I will leave `switch` as I think this will later be cleaner when I add `VOLUME_CONTROL` for example.

jfischer-phytec-iot replied:
This switch block will not cleaner then, please move it to `handle_fu_req_mute`. switch{switch{if{if}}} is ugly

emob-nordic replied:
Are You suggesting this?
```
if (cs != USB_AUDIO_FU_MUTE_CONTROL) {
    return -ENOSUP;
}

return handle_fu_req_mute(...); 
```

jfischer-phytec-iot replied:
You can leave the first switch block, just move the mute control to handle_fu_req_mute(...);

emob-nordic commented:
Updated with @jfischer-phytec-iot comments (please revisit) and rebased.

jfischer-phytec-iot commented on samples/subsys/usb/audio/headphones_microphone/README.rst:
```diff
@@ -0,0 +1,37 @@
+.. _usb_audio_headphones_microphone:
+
+USB Audio Sample Application
+###################################
```
USB Audio Sample Application
#########################

jfischer-phytec-iot commented on samples/subsys/usb/audio/headphones_microphone/README.rst:
```diff
@@ -0,0 +1,37 @@
+describe the device.
+
+After you have built and flashed the sample app image to your board, plug the
+board into a host device, for example, a PC running Windows.
```
Window OS or better does not claim on a specific OS as it should work on Linux Host, Window OS, Mac OS....

jfischer-phytec-iot commented on samples/subsys/usb/audio/headset/README.rst:
```diff
@@ -0,0 +1,36 @@
+.. _usb_audio_headset:
+
+USB Audio Sample Application
+###################################
```
USB Audio Sample Application
#########################

jfischer-phytec-iot commented on samples/subsys/usb/audio/headset/README.rst:
```diff
@@ -0,0 +1,36 @@
+describe the device.
+
+After you have built and flashed the sample app image to your board, plug the
+board into a host device, for example, a PC running Windows.
```
here the same about Windows

carlescufi reviewed (approved):
Very nice work Emil, thanks!

arvid-r commented:
Well done! Great work.
